server/user.go

```diff
@@ -0,0 +1,284 @@
+		return appErr
+	}
+	if appErr := p.API.KVSet(tokenKeyByRemoteID+info.RemoteID, data); appErr != nil {
+		return appErr
+	}
+	return nil
+}
+
+func (p *Plugin) GetUserInfo(userID string) (*UserInfo, error) {
+	infoBytes, appErr := p.API.KVGet(tokenKey + userID)
+	if appErr != nil || infoBytes == nil {
+		return nil, errors.New("must connect user account to Microsoft first")
+	}
+
+	key := []byte(p.getConfiguration().EncryptionKey)
+	return DecryptUserInfo(infoBytes, key)
+}
+
+func (p *Plugin) RemoveUser(userID string) error {
+	info, err := p.GetUserInfo(userID)
+	if err != nil {
+		return err
+	}
+
+	errByMattermostID := p.API.KVDelete(tokenKey + userID)
+	errByRemoteID := p.API.KVDelete(tokenKeyByRemoteID + info.RemoteID)
+	if errByMattermostID != nil {
+		return errByMattermostID
+	}
+	if errByRemoteID != nil {
+		return errByRemoteID
+	}
+	return nil
+}
+
+func encrypt(key, data []byte) (string, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create a cipher block, check key")
+	}
+
+	data = pad(data)
+	ciphertext := make([]byte, aes.BlockSize+len(data))
+	iv := ciphertext[:aes.BlockSize]
+	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
+		return "", errors.Wrap(err, "readFull was unsuccessful, check buffer size")
+	}
+
+	cfb := cipher.NewCFBEncrypter(block, iv)
+	cfb.XORKeyStream(ciphertext[aes.BlockSize:], data)
+	finalMsg := base64.URLEncoding.EncodeToString(ciphertext)
+	return finalMsg, nil
+}
+
+func decrypt(key []byte, text string) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create a cipher block, check key")
+	}
+
+	decodedMsg, err := base64.URLEncoding.DecodeString(text)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not decode the message")
+	}
+
+	if (len(decodedMsg) % aes.BlockSize) != 0 {
+		return nil, errors.New("blocksize must be multiple of decoded message length")
+	}
+
+	iv := decodedMsg[:aes.BlockSize]
+	msg := decodedMsg[aes.BlockSize:]
+
+	cfb := cipher.NewCFBDecrypter(block, iv)
+	cfb.XORKeyStream(msg, msg)
+
+	unpadMsg, err := unpad(msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "unpad error, check key")
+	}
+
+	return unpadMsg, nil
+}
+
+func pad(src []byte) []byte {
+	padding := aes.BlockSize - len(src)%aes.BlockSize
+	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
+	return append(src, padtext...)
+}
+
+func unpad(src []byte) ([]byte, error) {
+	length := len(src)
+	unpadding := int(src[length-1])
+
+	if unpadding > length {
+		return nil, errors.New("unpad error. This could happen when incorrect encryption key is used")
+	}
+
+	return src[:(length - unpadding)], nil
+}
+
+func generateSecret() (string, error) {
+	b := make([]byte, 256)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+	s := base64.RawStdEncoding.EncodeToString(b)
+
+	s = s[:32]
+
+	return s, nil
+}
+
+const pageSize = 100
+
+func (p *Plugin) resetAllOAuthTokens() error {
```

Comment:
Just so I understand.
This gets all the users and remove the token. Which (if I am not mistaken) will kick off again the connection flow, which will give us a whole new user info. Am I right? Or did I miss something?

If that is right, what is the difference between doing this, and doing a KVDeleteAll?

Reply:
Thanks for the suggestion. Verified that the KV is not used for anything else.

Reply:
> Verified that the KV is not used for anything else.

It _may_ be used for something in the future though. 2/5 to avoid an operation that is potentially destructive later